text/0000-transparent-unions.md

```diff
@@ -0,0 +1,229 @@
+- Feature Name: transparent_enunions
+- Start Date: 2019-02-13
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Allow `#[repr(transparent)]` on `union`s and univariant `enum`s that have exactly one non-zero-sized field (just like `struct`s).
+
+# Motivation
+[motivation]: #motivation
+
+Some `union` types are thin newtype-style wrappers around another type, like `MaybeUninit<T>` (and [once upon a time](https://doc.rust-lang.org/1.28.0/src/core/mem.rs.html#955), `ManuallyDrop<T>`). This type is intended to be used in the same places as `T`, but without being `#[repr(transparent)]` the actual compatibility between it and `T` is left unspecified.
+
+Likewise, some `enum` types only have a single variant, and are similarly thin wrappers around another type.
+
+Making types like these `#[repr(transparent)]` would be useful in certain cases. For example, making the type `Wrapper<T>` (which is a `union` or univariant `enum` with a single field of type `T`) transparent:
+
+- Clearly expresses the intent of the developer.
+- Protects against accidental violations of that intent (e.g., adding a new variant or non-ZST field will result in a compiler error).
```

Comment:
I'm not sure there's much to explore here since this is pretty different from `repr(transparent)`. `MyUnion` needs to be `repr(C)` (if not `repr(transparent)`) to be FFI-safe, but the developer can add additional sized fields to `MyUnion` (something that `repr(transparent)` would have prevented) (and there's no good way to prevent additional sized fields from being added to a `union`).

Reply:
Indeed. Is there an easy way to write a compile-time test to check that a union is indeed transparent ? If not, that's an argument that might be worth mentioning (e.g. without explicit transparent there is no easy way to write a test to check it).